api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,616 @@
+	var resourceConsumers []scaler.ResourceConsumer
+
+	serviceARNs, err := e.getServiceARNsForCluster(clusterName)
+	if err != nil {
+		return nil, err
+	}
+
+	services, err := e.getServicesFromServiceARNs(clusterName, serviceARNs)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, service := range services {
+		deployIDCopies := map[string]int{}
+		// deployment.RunningCount is the number of containers already running on an instance
+		// deployment.PendingCount is the number of containers that are alraedy on an instance, but are being pulled
+		// we only care about containers that are not on instances yet
+		for _, d := range service.Deployments {
+			desiredCount := aws.Int64Value(d.DesiredCount)
+			runningCount := aws.Int64Value(d.RunningCount)
+			pendingCount := aws.Int64Value(d.PendingCount)
+			if numPending := desiredCount - (runningCount + pendingCount); numPending > 0 {
+				deployIDCopies[aws.StringValue(d.TaskDefinition)] = int(numPending)
+			}
+		}
+
+		if len(deployIDCopies) == 0 {
+			continue
+		}
+
+		// iterate through deploys and their copies
+		for deployID, numCopies := range deployIDCopies {
+			for i := 0; i < numCopies; i++ {
+				consumers, err := e.getContainerResourceFromDeploy(deployID)
+				if err != nil {
+					return nil, err
+				}
+
+				resourceConsumers = append(resourceConsumers, consumers...)
+			}
+		}
+	}
+
+	return resourceConsumers, nil
+}
+
+func (e *EnvironmentScaler) getJobTaskResourceConsumers(clusterName string) ([]scaler.ResourceConsumer, error) {
+	var resourceConsumers []scaler.ResourceConsumer
+	jobs, err := e.JobStore.SelectAll()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, job := range jobs {
+		if job.Type == models.CreateTaskJob {
+			// Don't check for Pending jobs.
+			// It's more efficient and cost-effective to only scale for jobs that are InProgress.
+			// Jobs could be stuck in the Pending state for hours waiting for a worker to pick them up.
+			if job.Status == models.InProgressJobStatus {
+				var req models.CreateTaskRequest
+				if err := json.Unmarshal([]byte(job.Request), &req); err != nil {
+					return nil, err
+				}
+
+				if req.EnvironmentID == clusterName {
+					taskResourceConsumers, err := e.getContainerResourceFromDeploy(req.DeployID)
+					if err != nil {
+						return nil, err
+					}
+
+					resourceConsumers = append(resourceConsumers, taskResourceConsumers...)
+				}
+			}
+		}
+	}
+
+	return resourceConsumers, nil
+}
+
+func (e *EnvironmentScaler) getResourceProviders(clusterName string) ([]*scaler.ResourceProvider, error) {
+	var result []*scaler.ResourceProvider
+
+	containerInstanceARNs, err := e.getActiveContainerInstanceARNsForCluster(clusterName)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(containerInstanceARNs) == 0 {
+		return result, nil
+	}
+
+	containerInstances, err := e.getContainerInstancesFromARNs(clusterName, containerInstanceARNs)
```

Comment:
either the naming of these functions is off, or something is very wrong. A container instance != cluster instance. We should only care about cluster instances in this context.

Reply:
Where are you drawing the line, exactly? According to AWS' docs, a "container instance" is, essentially, a "cluster instance": http://docs.aws.amazon.com/AmazonECS/latest/developerguide/ECS_instances.html

Reply:
Ah, I see I was mistaken. I assumed `ContainerInstance` represented a single docker container. I'm gonna 100% blame AWS for that one though, horrible naming.